File: java/linkedlist/RemoveNthNodeFromEndOfList.java
package linkedlist;

public class RemoveNthNodeFromEndOfList {
    /* debug还得靠自己！基本思路非常简单 要从后面往前数k个数可以用快慢指针
     * 这里有几个细节 首先我们不需要后面往前k个数 而是要k+1 因为我们要用next = next.next delete
     * 其次就是为什么slow和fast都要从dummy开始呢 具体我也说不清 反正这样能过1,2,3,4,5 5的testcase
     * 正好把1 remove掉 从head开始就不可能把head remove 然后k的位置也happen to be正确
     */
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode slow = new ListNode(), fast = new ListNode();
        slow.next = head;
        fast.next = head;
        ListNode output = slow;
        for(int i = 0;fast != null && i < n + 1;i++) {
            fast = fast.next;
        }
        while(fast != null){
            slow = slow.next;
            fast = fast.next;
        }
        slow.next = slow.next.next;
        return output.next;
    }
}

File: c++/linkedlist/removeNthNodeFromEndOfList.cpp
#include <include/listnode.h>
/**
 * 19. Remove Nth Node From End of List
 * Arrive at the same solution as Java version
 * 这里并不实际上是快慢指针 而是有一个指针先跑n步 然后一起跑 第二个指针就停在了我们想要删除的地方
 * 这里需要一些细致的推算 为什么一定要从dummy开始是因为如果我们要删除第一个element 那么无论如何没法从前面删掉head
 * 只能直接return 但是这样的方法会在后面的testcase遇到问题 所以还是dummy起步 first多跑一步就行了
*/
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *first = new ListNode(0), *second = new ListNode(0), *dummy=second;
        first->next = head;
        second->next = head;
        for(int i = 0;i < n+1;i++) first = first->next;
        // if(!first) return nullptr;
        while(first != nullptr) {
            first = first->next;
            second = second->next;
        }
        // cout << second->val << endl;
        second->next = second->next->next;
        return dummy->next;
    }
};